File: csharp/lib/lim/Visitors/VisitorFilter.cs
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2017 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

using System;
using System.Text;

namespace Columbus.Lim.Asg.Visitors {

    /// <summary>
    /// VisitorFilter class filters out nodes from the ASG.
    /// </summary>
#if (LIM_INTERNAL)
    internal
#else
    public
#endif
    class VisitorFilter : Visitor {

        /// <summary>Pointer type to the set(Not)FilteredThisNodeOnly() method.</summary>
        protected delegate void SelectorType(uint id);

        /// <summary>Pointer to the set(Not)FilteredThisNodeOnly() method.</summary>
        protected SelectorType selector;

        /// <summary>
        /// Constructor for VisitorFilter.
        /// </summary>
        /// <param name="factory">[in] The factory which contains the ASG.</param>
        /// <param name="filter">[in] If this parameter is true then the visitor will filter the nodes out. If it is false then it set the nodes to not filtered.</param>
        public VisitorFilter(Factory factory, bool filter) {
            if (filter)
                selector = new SelectorType(factory.setFilteredThisNodeOnly);
            else
                selector = new SelectorType(factory.setNotFilteredThisNodeOnly);
        }

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Base.Comment node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Base.Component node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Base.ControlFlowBlock node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.Attribute node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.AttributeAccess node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.Class node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.ClassGeneric node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.ClassGenericInstance node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.ClassGenericSpec node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.Friendship node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.GenericParameter node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.Method node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.MethodCall node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.MethodGeneric node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.MethodGenericInstance node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.MethodGenericSpec node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.Package node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Logical.Parameter node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Physical.File node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Physical.FileSystem node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Physical.Folder node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Type.SimpleType node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Type.Type node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Type.TypeFormerArray node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Type.TypeFormerMethod node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Type.TypeFormerNonType node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Type.TypeFormerPointer node, bool callVirtualBase) { selector(node.Id); } 

        /// <summary>
        /// Deselects the actual node (and only this).
        /// </summary>
        /// <param name="node">[in] The node which is visited.</param>
        public override void visitEnd(Columbus.Lim.Asg.Nodes.Type.TypeFormerType node, bool callVirtualBase) { selector(node.Id); } 

    }

}

File: lib/lim/src/visitors/VisitorFilter.cpp
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2018 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

#include "lim/inc/lim.h"


namespace columbus { namespace lim { namespace asg {
VisitorFilter::VisitorFilter(bool filter) : Visitor() ,selector( filter?&Factory::setFilteredThisNodeOnly : &Factory::setNotFilteredThisNodeOnly) {
}

VisitorFilter::~VisitorFilter() {
}

void VisitorFilter::visitEnd(const base::Comment& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const base::Component& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const base::ControlFlowBlock& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::Attribute& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::AttributeAccess& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::Class& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::ClassGeneric& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::ClassGenericInstance& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::ClassGenericSpec& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::Friendship& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::GenericParameter& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::Method& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::MethodCall& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::MethodGeneric& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::MethodGenericInstance& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::MethodGenericSpec& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::Package& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const logical::Parameter& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const physical::File& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const physical::FileSystem& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const physical::Folder& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const type::SimpleType& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const type::Type& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const type::TypeFormerArray& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const type::TypeFormerMethod& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const type::TypeFormerNonType& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const type::TypeFormerPointer& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}

void VisitorFilter::visitEnd(const type::TypeFormerType& node , bool callVirtualBase /*= true*/) {
  (node.getFactory().*selector)(node.getId());
}


}}}
